File: backend/internal/service/folder_service.go
package service

import (
	"context"
	"fmt"

	"connectrpc.com/connect"
	"github.com/oklog/ulid/v2"
	"google.golang.org/protobuf/types/known/emptypb"

	pb "github.com/plexreader/plexreader/backend/gen/plexreader/v1"
	"github.com/plexreader/plexreader/backend/gen/plexreader/v1/plexreaderv1connect"
	"github.com/plexreader/plexreader/backend/internal/storage"
)

// FolderService implements plexreaderv1connect.FolderServiceHandler.
type FolderService struct {
	store storage.FolderStore
}

func NewFolderService(store storage.FolderStore) plexreaderv1connect.FolderServiceHandler {
	return &FolderService{store: store}
}

func (s *FolderService) CreateFolder(ctx context.Context, req *connect.Request[pb.CreateFolderRequest]) (*connect.Response[pb.Folder], error) {
	f := req.Msg.Folder
	if f == nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("folder is required"))
	}
	folder := &storage.Folder{
		ID:       ulid.Make().String(),
		Name:     f.Name,
		ParentID: f.ParentId,
	}
	created, err := s.store.Create(ctx, folder)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return connect.NewResponse(folderToProto(created, 0)), nil
}

func (s *FolderService) GetFolder(ctx context.Context, req *connect.Request[pb.GetFolderRequest]) (*connect.Response[pb.Folder], error) {
	f, err := s.store.Get(ctx, req.Msg.Id)
	if err != nil {
		return nil, notFoundOrInternal(err)
	}
	counts, _ := s.store.GetUnreadCounts(ctx)
	return connect.NewResponse(folderToProto(f, int32(counts[f.ID]))), nil
}

func (s *FolderService) ListFolders(ctx context.Context, req *connect.Request[pb.ListFoldersRequest]) (*connect.Response[pb.ListFoldersResponse], error) {
	folders, err := s.store.List(ctx)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	counts, _ := s.store.GetUnreadCounts(ctx)
	protos := make([]*pb.Folder, len(folders))
	for i, f := range folders {
		protos[i] = folderToProto(f, int32(counts[f.ID]))
	}
	return connect.NewResponse(&pb.ListFoldersResponse{Folders: protos}), nil
}

func (s *FolderService) UpdateFolder(ctx context.Context, req *connect.Request[pb.UpdateFolderRequest]) (*connect.Response[pb.Folder], error) {
	f := req.Msg.Folder
	if f == nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("folder is required"))
	}
	updates := map[string]interface{}{
		"name":      f.Name,
		"parent_id": f.ParentId,
		"position":  f.Position,
	}
	updated, err := s.store.Update(ctx, f.Id, updates)
	if err != nil {
		return nil, notFoundOrInternal(err)
	}
	counts, _ := s.store.GetUnreadCounts(ctx)
	return connect.NewResponse(folderToProto(updated, int32(counts[updated.ID]))), nil
}

func (s *FolderService) DeleteFolder(ctx context.Context, req *connect.Request[pb.DeleteFolderRequest]) (*connect.Response[emptypb.Empty], error) {
	if err := s.store.Delete(ctx, req.Msg.Id); err != nil {
		return nil, notFoundOrInternal(err)
	}
	return connect.NewResponse(&emptypb.Empty{}), nil
}

func (s *FolderService) ReorderFolders(ctx context.Context, req *connect.Request[pb.ReorderFoldersRequest]) (*connect.Response[pb.ReorderFoldersResponse], error) {
	folders, err := s.store.Reorder(ctx, req.Msg.FolderIds)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	protos := make([]*pb.Folder, len(folders))
	for i, f := range folders {
		protos[i] = folderToProto(f, 0)
	}
	return connect.NewResponse(&pb.ReorderFoldersResponse{Folders: protos}), nil
}

File: ui/lib/services/folder_service.dart
// SPDX-License-Identifier: LGPL-2.1-or-later
import '../models/folder.dart';
import 'api_client.dart';

class FolderService {
  static const _service = 'plexreader.v1.FolderService';
  final ApiClient _client;

  FolderService(this._client);

  Future<List<Folder>> listFolders() async {
    final res = await _client.post(_service, 'ListFolders', {});
    final list = res['folders'] as List<dynamic>? ?? [];
    return list.map((e) => Folder.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Folder> createFolder(String name, {String? parentId, int? position}) async {
    final res = await _client.post(_service, 'CreateFolder', {
      'folder': {
        'name': name,
        if (parentId != null) 'parentId': parentId,
        if (position != null) 'position': position,
      },
    });
    return Folder.fromJson(res);
  }

  Future<Folder> updateFolder(Folder folder) async {
    final res = await _client.post(_service, 'UpdateFolder', {
      'folder': folder.toJson(),
    });
    return Folder.fromJson(res);
  }

  Future<void> deleteFolder(String id) async {
    await _client.post(_service, 'DeleteFolder', {'id': id});
  }

  Future<List<Folder>> reorderFolders(List<String> folderIds) async {
    final res = await _client.post(_service, 'ReorderFolders', {
      'folderIds': folderIds,
    });
    final list = res['folders'] as List<dynamic>? ?? [];
    return list.map((e) => Folder.fromJson(e as Map<String, dynamic>)).toList();
  }
}
